src/mem/metaslab.h

```diff
@@ -1,20 +1,408 @@
 #pragma once
 
-#include "../backend/metatypes.h"
+#include "../backend/backend_concept.h"
 #include "../ds/helpers.h"
 #include "../ds/seqset.h"
-#include "../mem/remoteallocator.h"
 #include "freelist.h"
 #include "sizeclasstable.h"
 
 namespace snmalloc
 {
-  // The Metaslab represent the status of a single slab.
-  class alignas(CACHELINE_SIZE) Metaslab
+  struct RemoteAllocator;
+
+  /**
+   * Remotes need to be aligned enough that the bottom bits have enough room for
+   * all the size classes, both large and small. An additional bit is required
+   * to separate backend uses.
+   */
+  static constexpr size_t REMOTE_MIN_ALIGN =
+    bits::max<size_t>(CACHELINE_SIZE, SIZECLASS_REP_SIZE) << 1;
+
+  /**
+   * Base class for the templated MetaEntry.  This exists to avoid needing a
+   * template parameter to access constants that are independent of the
+   * template parameter.
+   */
+  class MetaEntryBase
```

Comment:
Given that `metaslab` is not a great name for what's going on here and nothing about `MetaEntry` directly references `MetaSlab` (because it's indirected through its template for the backend), could this and its child be moved out to its own `src/mem/metaentry.h`?

Reply:
I have rearranged this so that:

 - `MetaEntry` is now `FrontendMetaEntry`.
 - `Metaslab` is now `FrontendSlabMetadata`.
 - `FrontendMetaEntry` is declared after `FrontendSlabMetadata` so that it can check that its template argument is a subclass of `FrontendSlabMetadata`.
 - `mem/metaslab.h` is now called `mem/metadata.h` and includes all three classes that describe metadata for the front end.